src/rt/backtrace/dwarf.d

```diff
@@ -0,0 +1,492 @@
+ * Source: $(DRUNTIMESRC src/rt/backtrace/dwarf.d)
+ */
+
+module rt.backtrace.dwarf;
+
+version(linux):
+
+import rt.util.container.array;
+import rt.backtrace.elf;
+
+import core.stdc.string : strlen, memchr;
+
+struct Location
+{
+    const(char)[] file = null; // file is missing directory, but DMD emits directory directly into file
+    int line = -1;
+    size_t address;
+}
+
+int traceHandlerOpApplyImpl(bool includeBacktraceSymbols)(const void*[] callstack, scope int delegate(ref size_t, ref const(char[])) dg)
+{
+    import core.stdc.stdio : snprintf;
+
+    Array!Location locations;
+
+    static if (includeBacktraceSymbols)
+    {
+        import core.sys.linux.execinfo : backtrace_symbols;
+        import core.sys.posix.stdlib : free;
+
+        const framelist = backtrace_symbols(callstack.ptr, cast(int) callstack.length);
+        scope(exit) free(cast(void*) framelist);
+    }
+
+    // find address -> file, line mapping using dwarf debug_line
+    ElfFile file;
+    ElfSection dbgSection;
+    if (ElfFile.openSelf(&file))
+    {
+        auto stringSectionHeader = ElfSectionHeader(&file, file.ehdr.e_shstrndx);
+        auto stringSection = ElfSection(&file, &stringSectionHeader);
+
+        auto dbgSectionIndex = findSectionByName(&file, &stringSection, ".debug_line");
+        if (dbgSectionIndex != -1)
+        {
+            auto dbgSectionHeader = ElfSectionHeader(&file, dbgSectionIndex);
+            dbgSection = ElfSection(&file, &dbgSectionHeader);
+            // debug_line section found and loaded
+
+            // resolve addresses
+            locations.length = callstack.length;
+            foreach(size_t i; 0 .. callstack.length)
+                locations[i].address = cast(size_t) callstack[i];
+
+            resolveAddresses(&dbgSection, locations[]);
+        }
+    }
+
+    int ret = 0;
+    foreach (size_t i; 0 .. callstack.length)
+    {
+        char[1536] buffer = void; buffer[0] = 0;
+        char[256] addressBuffer = void; addressBuffer[0] = 0;
+
+        if (locations.length > 0 && locations[i].line != -1)
+            snprintf(addressBuffer.ptr, addressBuffer.length, "%.*s:%d ", cast(int) locations[i].file.length, locations[i].file.ptr, locations[i].line);
+        else
+            addressBuffer[] = "??:? \0";
+
+        static if (includeBacktraceSymbols)
+        {
+            char[1024] symbolBuffer = void;
+            auto symbol = getDemangledSymbol(framelist[i][0 .. strlen(framelist[i])], symbolBuffer);
+            if (symbol.length > 0)
+                snprintf(buffer.ptr, buffer.length, "%s%.*s [0x%x]", addressBuffer.ptr, cast(int) symbol.length, symbol.ptr, callstack[i]);
+            else
+                snprintf(buffer.ptr, buffer.length, "%s[0x%x]", addressBuffer.ptr, callstack[i]);
+        }
+        else
+        {
+            snprintf(buffer.ptr, buffer.length, "%s[0x%x]", addressBuffer.ptr, callstack[i]);
+        }
+
+        auto output = buffer[0 .. strlen(buffer.ptr)];
```

Comment:
This could probably use the return value of `snprintf` instead of searching for the null terminator?